examples/echo/echo_requester.cpp

```diff
@@ -0,0 +1,250 @@
+ *      CHIP Echo Protocol.
+ *
+ *      The CHIP Echo Protocol implements two simple methods, in the
+ *      style of ICMP ECHO REQUEST and ECHO REPLY, in which a sent
+ *      payload is turned around by the responder and echoed back to
+ *      the originator.
+ *
+ */
+
+#include "common.h"
+
+#include <protocols/echo/Echo.h>
+#include <support/ErrorStr.h>
+
+#define ECHO_CLIENT_PORT (CHIP_PORT + 1)
+
+namespace {
+
+// Max value for the number of EchoRequests sent.
+constexpr size_t kMaxEchoCount = 3;
+
+// The CHIP Echo interval time in microseconds.
+constexpr int32_t gEchoInterval = 1000000;
+
+// The EchoClient object.
+Protocols::EchoClient gEchoClient;
+
+SecureSessionMgr<Transport::UDP> gSessionManager;
+
+Inet::IPAddress gDestAddr;
+
+// The last time a CHIP Echo was attempted to be sent.
+uint64_t gLastEchoTime = 0;
+
+// True, if the EchoClient is waiting for an EchoResponse
+// after sending an EchoRequest, false otherwise.
+bool gWaitingForEchoResp = false;
+
+// Count of the number of EchoRequests sent.
+uint64_t gEchoCount = 0;
+
+// Count of the number of EchoResponses received.
+uint64_t gEchoRespCount = 0;
+
+uint64_t Now(void)
+{
+    struct timeval now;
+    gettimeofday(&now, NULL);
+
+    return ((uint64_t) now.tv_sec * 1000000) + (uint64_t) now.tv_usec;
+}
+
+bool EchoIntervalExpired(void)
+{
+    return (Now() >= gLastEchoTime + gEchoInterval);
+}
+
+void FormulateEchoRequestBuffer(System::PacketBuffer *& payloadBuf)
+{
+    payloadBuf = System::PacketBuffer::New();
+    if (payloadBuf == NULL)
+    {
+        printf("Unable to allocate PacketBuffer\n");
+        return;
+    }
+
+    // Add some application payload data in the buffer.
+    char * p    = (char *) payloadBuf->Start();
+    int32_t len = sprintf(p, "Echo Message %" PRIu64 "\n", gEchoCount);
+
+    // Set the datalength in the buffer appropriately.
+    payloadBuf->SetDataLength((uint16_t) len);
+}
+
+CHIP_ERROR SendEchoRequest(void)
+{
+    CHIP_ERROR err                    = CHIP_NO_ERROR;
+    System::PacketBuffer * payloadBuf = NULL;
+
+    gLastEchoTime = Now();
+
+    FormulateEchoRequestBuffer(payloadBuf);
+    if (payloadBuf == NULL)
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    printf("\nSend echo request message to Node: %lu\n", kServerDeviceId);
+
+    err = gEchoClient.SendEchoRequest(kServerDeviceId, payloadBuf);
+
+    // Set the local buffer to NULL after passing it down to
+    // the lower layers who are now responsible for freeing
+    // the buffer.
+    payloadBuf = NULL;
+
+    if (err == CHIP_NO_ERROR)
+    {
+        gWaitingForEchoResp = true;
+        gEchoCount++;
+    }
+    else
+    {
+        printf("Send echo request failed, err: %s\n", ErrorStr(err));
+    }
+
+    return err;
+}
+
+CHIP_ERROR EstablishSecureSession()
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    SecurePairingUsingTestSecret * testSecurePairingSecret = Platform::New<SecurePairingUsingTestSecret>(
```

Comment:
Will the example soon move to using a real unicast secure channel using CASE? PASE is only meant for commissioning.

Reply:
I believe so, right now, this is the only option.